non-critical-infra/modules/matrix-synapse.nix

```diff
@@ -0,0 +1,112 @@
+{ config
+, pkgs
+, ...
+}:
+
+{
+  imports = [
+    ./nginx.nix
+    ./postgresql.nix
+  ];
+
+  fileSystems."/var/lib/matrix-synapse" = {
+    device = "zroot/root/matrix-synapse";
+    fsType = "zfs";
+    options = [
+      "zfsutil"
+    ];
+  };
+
+  services.postgresql = {
+    ensureUsers = [ {
+      name = "matrix-synapse";
+      ensureDBOwnership = true;
+    } ];
+    ensureDatabases = [
+      "matrix-synapse"
+    ];
+  };
+
+  services.redis.servers.matrix-synapse = {
+    enable = true;
+  };
+
+  environment.systemPackages = with pkgs; [
+    matrix-synapse-tools.synadm
+  ];
+
+  # create /run/matrix-synapse for the unix socket listener
+  systemd.services.matrix-synapse.ServiceConfing.RuntimeDirectory = "matrix-synapse";
+
+  services.matrix-synapse = {
+    enable = true;
+    withJemalloc = true;
+
+    # https://github.com/element-hq/synapse/blob/master/docs/usage/configuration/config_documentation.md
+    settings = {
+      server_name = "nixos.org";
+      public_base = "https://matrix.nixos.org";
+      admin_contact = "infra@nixos.org";
+
+      allow_public_rooms_over_federation = true;
+      allow_public_rooms_without_auth = true;
+
+      max_upload_size = "50M";
+
```

Comment:
It might be required to enable `experimental_features.spaces_enabled = true`, but maybe it's stable now.

Reply:
> - Remove the experimental `spaces_enabled` flag. The spaces features are always available now.

Removed in 2021. https://github.com/element-hq/synapse/blob/232adfb95f9ec726c2c09f3329c098aa9060c1f7/docs/changelogs/CHANGES-2021.md?plain=1#L1428